File: src/main/java/com/example/gatewayservice/filter/AuthorizationHeaderFilter.java
package com.example.gatewayservice.filter;

import static org.springframework.http.HttpHeaders.AUTHORIZATION;
import static org.springframework.http.HttpStatus.UNAUTHORIZED;

import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cloud.gateway.filter.GatewayFilter;
import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.security.oauth2.jwt.JwtException;
import org.springframework.security.oauth2.jwt.ReactiveJwtDecoder;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;

/**
 * JWKS 로 서명·만료·발급자를 검증하고, 라우트가 요구하는 role 과 audience(어느 앱의 토큰인지)를 확인한다.
 * 통과하면 X-Auth-User-Id(sub) 와 X-Auth-Client-Id(aud) 를 넣어 준다.
 */
@Component
@Slf4j
public class AuthorizationHeaderFilter extends AbstractGatewayFilterFactory<AuthorizationHeaderFilter.Config> {

    public static final String USER_ID_HEADER = "X-Auth-User-Id";
    public static final String CLIENT_ID_HEADER = "X-Auth-Client-Id";

    private final ReactiveJwtDecoder jwtDecoder;

    public AuthorizationHeaderFilter(ReactiveJwtDecoder jwtDecoder) {
        super(Config.class);
        this.jwtDecoder = jwtDecoder;
    }

    /** yml 에서 `AuthorizationHeaderFilter=ROLE_ADMIN,modu-admin` 처럼 role, audience 순으로 받는다. */
    @Override
    public List<String> shortcutFieldOrder() {
        return List.of("role", "audience");
    }

    @Override
    public GatewayFilter apply(Config config) {
        return (exchange, chain) -> {
            String header = exchange.getRequest().getHeaders().getFirst(AUTHORIZATION);
            if (header == null || !header.startsWith("Bearer ")) {
                return onError(exchange, "No authorization header", UNAUTHORIZED);
            }
            return jwtDecoder.decode(header.substring("Bearer ".length()))
                    .flatMap(jwt -> {
                        if (config.getAudience() != null && !config.getAudience().isBlank()
                                && (jwt.getAudience() == null || !jwt.getAudience().contains(config.getAudience()))) {
                            return onError(exchange, "JWT audience mismatch: requires " + config.getAudience(), UNAUTHORIZED);
                        }
                        List<String> roles = jwt.getClaimAsStringList("roles");
                        if (roles == null || !roles.contains(config.getRole())) {
                            return onError(exchange, "JWT role mismatch: requires " + config.getRole(), UNAUTHORIZED);
                        }
                        String userId = jwt.getSubject();
                        String clientId = jwt.getAudience() == null || jwt.getAudience().isEmpty() ? "" : jwt.getAudience().get(0);
                        ServerHttpRequest.Builder mutated = exchange.getRequest().mutate().header(CLIENT_ID_HEADER, clientId);
                        if (userId != null && !userId.isBlank()) {
                            mutated.header(USER_ID_HEADER, userId);
                        }
                        return chain.filter(exchange.mutate().request(mutated.build()).build());
                    })
                    .onErrorResume(JwtException.class, e -> onError(exchange, "JWT token is not valid: " + e.getMessage(), UNAUTHORIZED));
        };
    }

    private Mono<Void> onError(ServerWebExchange exchange, String err, HttpStatus httpStatus) {
        ServerHttpResponse response = exchange.getResponse();
        response.setStatusCode(httpStatus);
        log.error(err);
        return response.setComplete();
    }

    public static class Config {
        private String role = "ROLE_USER";
        private String audience;
        public String getRole() { return role; }
        public void setRole(String role) { this.role = role; }
        public String getAudience() { return audience; }
        public void setAudience(String audience) { this.audience = audience; }
    }
}

File: gateway-service/src/main/kotlin/com/example/gatewayservice/filter/AuthorizationHeaderFilter.kt
package com.example.gatewayservice.filter

import org.slf4j.LoggerFactory
import org.springframework.cloud.gateway.filter.GatewayFilter
import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.security.oauth2.jwt.JwtException
import org.springframework.security.oauth2.jwt.ReactiveJwtDecoder
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono

/**
 * JWKS 로 서명·만료·발급자를 검증하고, 라우트가 요구하는 role 과 audience(어느 앱의 토큰인지)를 확인한다.
 * 통과하면 X-Auth-User-Id(sub) 와 X-Auth-Client-Id(aud) 를 넣어 준다.
 */
@Component
class AuthorizationHeaderFilter(
    private val jwtDecoder: ReactiveJwtDecoder,
) : AbstractGatewayFilterFactory<AuthorizationHeaderFilter.Config>(Config::class.java) {

    companion object {
        const val USER_ID_HEADER = "X-Auth-User-Id"
        const val CLIENT_ID_HEADER = "X-Auth-Client-Id"
        private const val BEARER = "Bearer "
    }

    private val log = LoggerFactory.getLogger(AuthorizationHeaderFilter::class.java)

    /** yml 에서 `AuthorizationHeaderFilter=ROLE_ADMIN,modu-admin` 처럼 role, audience 순으로 받는다. */
    override fun shortcutFieldOrder(): List<String> = listOf("role", "audience")

    override fun apply(config: Config): GatewayFilter =
        GatewayFilter { exchange, chain ->
            val header = exchange.request.headers.getFirst(HttpHeaders.AUTHORIZATION)
            if (header == null || !header.startsWith(BEARER)) {
                return@GatewayFilter onError(exchange, "No authorization header", HttpStatus.UNAUTHORIZED)
            }
            jwtDecoder.decode(header.substring(BEARER.length))
                .flatMap { jwt ->
                    val requiredAudience = config.audience
                    if (!requiredAudience.isNullOrBlank() && jwt.audience?.contains(requiredAudience) != true) {
                        return@flatMap onError(exchange, "JWT audience mismatch: requires $requiredAudience", HttpStatus.UNAUTHORIZED)
                    }
                    val roles = jwt.getClaimAsStringList("roles")
                    if (roles == null || !roles.contains(config.role)) {
                        return@flatMap onError(exchange, "JWT role mismatch: requires ${config.role}", HttpStatus.UNAUTHORIZED)
                    }
                    val userId = jwt.subject
                    val clientId = jwt.audience?.firstOrNull() ?: ""
                    val mutated = exchange.request.mutate().header(CLIENT_ID_HEADER, clientId)
                    if (!userId.isNullOrBlank()) {
                        mutated.header(USER_ID_HEADER, userId)
                    }
                    chain.filter(exchange.mutate().request(mutated.build()).build())
                }
                .onErrorResume(JwtException::class.java) { e ->
                    onError(exchange, "JWT token is not valid: ${e.message}", HttpStatus.UNAUTHORIZED)
                }
        }

    private fun onError(exchange: ServerWebExchange, err: String, httpStatus: HttpStatus): Mono<Void> {
        val response = exchange.response
        response.statusCode = httpStatus
        log.error(err)
        return response.setComplete()
    }

    /** 라우트 인자. 기본 role 은 ROLE_USER, audience 를 비우면 어느 앱의 토큰이든 받는다. */
    class Config {
        var role: String = "ROLE_USER"
        var audience: String? = null
    }
}
